lib/Targets/AIETargetAirbin.cpp

```diff
@@ -0,0 +1,1746 @@
+static constexpr auto TILE_ADDR_ROW_WIDTH = 5u;
+
+static constexpr auto TILE_ADDR_COL_SHIFT =
+    TILE_ADDR_ROW_SHIFT + TILE_ADDR_ROW_WIDTH;
+static constexpr auto TILE_ADDR_COL_WIDTH = 7u;
+
+static constexpr auto TILE_ADDR_ARR_SHIFT =
+    TILE_ADDR_COL_SHIFT + TILE_ADDR_COL_WIDTH;
+
+/*
+ * Tile address format:
+ * --------------------------------------------
+ * |                7 bits  5 bits   18 bits  |
+ * --------------------------------------------
+ * | Array offset | Column | Row | Tile addr  |
+ * --------------------------------------------
+ */
+class TileAddress {
+public:
+  TileAddress(uint8_t column, uint8_t row, uint64_t array_offset = 0x800u)
+      : array_offset{array_offset}, column{column}, row{row} {}
+
+  // SFINAE is used here to choose the copy constructor for `TileAddress`,
+  // and this constructor for all other classes.
+  template <
+      typename Op,
+      std::enable_if_t<not std::is_same<Op, TileAddress>::value, bool> = true>
+  TileAddress(Op &op)
+      : TileAddress{static_cast<uint8_t>(op.colIndex()),
+                    static_cast<uint8_t>(op.rowIndex())} {}
+
+  uint64_t fullAddress(uint32_t register_offset) const {
+    return (array_offset << TILE_ADDR_ARR_SHIFT) |
+           (static_cast<uint64_t>(column) << TILE_ADDR_COL_SHIFT) |
+           (static_cast<uint64_t>(row) << TILE_ADDR_ROW_SHIFT) |
+           register_offset;
+  }
+
+  bool isShim() const { return row == 0; }
+
+  operator uint16_t() const {
+    return (static_cast<uint16_t>(column) << TILE_ADDR_ROW_WIDTH) | row;
+  }
+
+  uint8_t col() const { return column; }
+
+private:
+  uint64_t array_offset : 34;
+  uint8_t column : TILE_ADDR_COL_WIDTH;
+  uint8_t row : TILE_ADDR_ROW_WIDTH;
+};
+
+static_assert(sizeof(TileAddress) <= sizeof(uint64_t),
+              "Tile addresses are at most 64-bits");
+
+class Address {
+public:
+  Address(TileAddress tile, uint32_t register_offset)
+      : tile{tile}, register_offset{register_offset} {}
+
+  operator uint64_t() const { return tile.fullAddress(register_offset); }
+
+  TileAddress destTile() const { return tile; }
+  uint32_t offset() const { return register_offset; }
+
+private:
+  TileAddress tile;
+  uint32_t register_offset : TILE_ADDR_ROW_SHIFT;
+};
+
+class Write {
+public:
+  Write(Address addr, uint32_t data) : address{addr}, data{data} {
+    assert(address % 4 == 0);
+  }
+
+  bool isJustBefore(const Write &rhs) const {
+    return static_cast<uint64_t>(this->address) + sizeof(data) ==
+           static_cast<uint64_t>(rhs.address);
+  }
+
+  uint64_t destination() const { return address; }
+  uint32_t relativeDest() const { return address.offset(); }
+
+  uint16_t tile() const { return address.destTile(); }
+  uint32_t value() const { return data; }
+
+  bool operator<(const Write &rhs) const {
+    return static_cast<uint64_t>(address) < static_cast<uint64_t>(rhs.address);
+  }
+
+private:
+  Address address;
+  uint32_t data;
+};
+
+static std::vector<Write> writes;
```

Comment:
Seems like all this static stuff should be a class, hiding the implementation of 'writes'?

Reply:
At the very least, it would seem that sorting this by write address would be a good thing?

Reply:
The writes are sorted at the end before generating the binary